nttt/tag_trimming.py

```diff
@@ -0,0 +1,109 @@
+import re
+
+# We are using a global variable because the function passed to re.sub() will only get the match as an argument.
+# We need another way of knowing the language so that we can enable language specific beahviour.
+t_lang = None
+
+# This list holds the locales identifying languages whose trailing hypens should be treated differently.
+hypen_langs = [
+    'nl-NL'
+]
+
+# These two lists contain characters that don't follow the general rules regarding formatting.
+permitted_end_chars = [
+    '>',
+    ')'
+]
+
+permitted_beg_chars = [
+    '('
+]
+
+# REs
+IN_TEXT_TAG = "\S\s*</?[\w\d]+>\s*\S"
+BEGINNING_TAG = "^<[\w\d]+>\s+\S"
+ENDING_TAG = "\S\s+</[\w\d]+>"
+
+def replacement_builder(matchobj):
+    """
+        Parameters:
+            - match -> A Match Object (see re's documentation) containing the matched string.
+
+        Returns:
+            A String containing the corrected version of the matched one.
+
+        This function crafts the correct tag from the one containing errors which we've found.
+        It's called by re.sub(). It distinguishes between the opening tags (i.e. <foo>) and closing
+        tags (i.e </foo>) so that we know how to fix things. It's also language aware so that we may
+        treat specific requirements separately like for example the Dutch trailing hypen which has
+        already been taken care of. Note we use tha auxiliary variable 'match' so as to avoid calling
+        the 'group()' method over and over again so as to reduce the call overhead.
+    """
+
+    match = matchobj.group(0)
+
+    # If we are dealing with a closing tag
+    if '/' in match:
```

Comment:
Despite all these comments the code is still hard to read because:
- the single method tries to deal with 3 different types of matches
- again we try to do more than was originally required